File: adapter/src/main/java/com/mdgd/pokemon/adapter/AbstractVH.java
package com.mdgd.pokemon.adapter;

import android.view.View;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public abstract class AbstractVH<ITEM> extends RecyclerView.ViewHolder {
    protected ITEM model = null;

    public AbstractVH(@NonNull View itemView) {
        super(itemView);
    }

    public void onViewAttachedToWindow() {

    }

    public void onViewDetachedFromWindow() {

    }

    protected boolean hasModel() {
        return model != null;
    }

    protected boolean bindSame() {
        return false;
    }

    public void onBindViewHolder(ITEM item) {
        if (!bindSame() && model == item) {
            return;
        }
        model = item;
        bind(item);
    }

    public void bind(ITEM item) {

    }
}

File: adapter/src/main/java/com/mdgd/pokemon/adapter/AbstractVH.kt
package com.mdgd.pokemon.adapter

import android.view.View
import androidx.recyclerview.widget.RecyclerView

abstract class AbstractVH<ITEM>(itemView: View) : RecyclerView.ViewHolder(itemView) {
    protected var model: ITEM? = null

    open fun onViewAttachedToWindow() {}

    open fun onViewDetachedFromWindow() {}

    protected fun hasModel(): Boolean = model != null

    protected open fun bindSame(): Boolean = false

    fun onBindViewHolder(item: ITEM) {
        if (!bindSame() && model === item) {
            return
        }
        model = item
        bind(item)
    }

    open fun bind(item: ITEM) {}
}
